src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -1195,17 +1463,735 @@ private static void ReportReadOnlyError(Symbol symbol, SyntaxNode node, BindValu
+                    // array elements and pointer dereferencing are readwrite variables
+                    return Binder.ExternalScope;
+
+                case BoundKind.RefValueOperator:
+                    // The undocumented __refvalue(tr, T) expression results in a variable of type T.
+                    // it is a readwrite variable, but could refer to unknown local data
+                    // we will constrain it to the current scope
+                    return scopeOfTheContainingExpression;
+
+                case BoundKind.DynamicMemberAccess:
+                case BoundKind.DynamicIndexerAccess:
+                    // dynamic expressions can be read and written to
+                    // can even be passed by reference (which is implemented via a temp)
+                    // it is not valid to escape them by reference though, so treat them as RValues here
+                    break;
+
+                case BoundKind.Parameter:
+                    var parameter = ((BoundParameter)expr).ParameterSymbol;
+
+                    // byval parameters can escape to method's top level.
+                    // others can be escape further, unless they are ref-like.
+                    // NOTE: "method" here means nearst containing method, lambda or nested method
+                    return parameter.RefKind == RefKind.None || parameter.Type?.IsByRefLikeType == true ?
+                        Binder.TopLevelScope :
+                        Binder.ExternalScope;
+
+                case BoundKind.Local:
+                    return ((BoundLocal)expr).LocalSymbol.RefEscapeScope;
+
+                case BoundKind.ThisReference:
+                    var thisref = (BoundThisReference)expr;
+
+                    // "this" is an RValue, unless in a struct.
+                    if (!thisref.Type.IsValueType)
+                    {
+                        break;
+                    }
+
+                    //"this" is not returnable by reference in a struct.
+                    // can ref escape to any other level
+                    return Binder.TopLevelScope;
+
+                case BoundKind.ConditionalOperator:
+                    var conditional = (BoundConditionalOperator)expr;
+
+                    // byref conditional defers to one operand. Since are here without errors, both operands must match.
+                    if (conditional.IsByRef)
+                    {
+                        return GetRefEscape(conditional.Consequence, scopeOfTheContainingExpression);
+                    }
+
+                    // otherwise it is an RValue
+                    break;
+
+                case BoundKind.FieldAccess:
+                    var fieldAccess = (BoundFieldAccess)expr;
+                    var fieldSymbol = fieldAccess.FieldSymbol;
+
+                    // fields that are static or belong to reference types can ref escape anywhere
+                    if (fieldSymbol.IsStatic || fieldSymbol.ContainingType.IsReferenceType)
```

Comment:
I didn't see the case for static field in the spec. Is it already covered? If not, lets update the spec.

Reply:
The static fields obviously have life times longer than the method that accesses them. If spec is vague about that, we should update.

---
In reply to: [135893591](https://github.com/dotnet/roslyn/pull/21334#discussion_r135893591) [](ancestors = 135893591)